Add option to support sources that taint existing, immutable objects.

wuestholz opened the pull request.

StevenArzt commented:
Please put the flag into the access path, not the abstraction. You should also define a proper semantics for the flag. What happens if I have an access path "a.b.c" with the flag set and then execute "x = a.b". Will the flag still be set on x.c afterwards? I'm not totally sure how a single flag can faithfully capture the semantics of access paths. Since not looking for string aliases is mainly an optimization, we could think about a global option to disable it - which would be more coarse-grained than a flag attached to the source and the access path, but it would cut short the semantic problem described above.

wuestholz commented:
@StevenArzt Thanks for looking at the changes. I have moved the flag to class 'AccessPath' and made it final.

About your question: Yes, at least based on my understanding of how the taint analysis operates. I have added a test based on your question. Please let me know if you were referring to something else or if you are worried about a different example.

I prefer not to have a global option for disabling this optimization since the optimization makes sense for regular sources. I would like to only disable it for select sources that correspond to user annotations.
